airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/MySqlCdcConnectorMetadataInjector.java

```diff
@@ -4,19 +4,30 @@
 
 package io.airbyte.integrations.source.mysql;
 
+import static io.airbyte.integrations.source.mysql.MySqlSource.CDC_DEFAULT_CURSOR;
 import static io.airbyte.integrations.source.mysql.MySqlSource.CDC_LOG_FILE;
 import static io.airbyte.integrations.source.mysql.MySqlSource.CDC_LOG_POS;
 
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.node.ObjectNode;
 import io.airbyte.integrations.debezium.CdcMetadataInjector;
+import java.time.Instant;
 
 public class MySqlCdcConnectorMetadataInjector implements CdcMetadataInjector {
 
+  private final Instant emittedAt;
+
+  public MySqlCdcConnectorMetadataInjector(final Instant emittedAt) {
+    this.emittedAt = emittedAt;
+  }
+
   @Override
   public void addMetaData(final ObjectNode event, final JsonNode source) {
+    final String cdcDefaultCursor =
+        String.format("%s_%s_%s", emittedAt.toString(), source.get("file").asText(), source.get("pos").asText());
```

Comment:
are all three of these values guaranteed to stringify to the same length? e.g. is it possible to end up with two records with cursors
```
123_x_y
4_x_y
```
where the second one _should_ sort lower (because 4 < 123), but we actually sort it higher (because `'4'` > `'1'`)

Reply:
+1 to verifying what Ed said. 

One other thing that I noticed while testing out CDC via PK is that the timestamp looks a little different based on whether debezium or Airbyte is populating the metadata fields. 

emitted_at for records emitted by Airbyte : `2023-07-27T17:14:38Z`
emitted_at for records emitted by Debezium : `2023-07-27T17:14:32.043504Z`

Of course - for this PR only Debezium is ever emitting records. Can we confirm that no matter the format, if one timestamp is before the other the cursor order will be respected? If not, we should file a ticket to make sure all records (whether emitted by Airbyte in the PK initial load sync, or DBZ on incremental runs) have the same ts structure

Reply:
@edgao, just verified with 2 different syncs and this is what the composite cursor looks like for those.
```
2023-08-01T20:54:18.167934430Z_mysql-bin.000108_4743280
2023-08-01T21:45:42.794296969Z_mysql-bin.000108_4930394
```
The timestamps all have 9 trailing digits. This is most likely the Debezium emitted at you mentioned @akashkulk.

Reply:
Sorry for the confusion here. Got confused between emitted_at and updated_at. This is all populated by a timestamp created by the Airbyte source connector at the start of the sync, so it should be the same format throughout.

Reply:
do file + pos also guarantee same string length across records? (this is super edgecasey)

Reply:
Verified that bin log file name ends at `999999` and the DB itself will crash if it ever gets to that point. As for the position, I couldn't verify that the number/format will stay the same length. However, I've just tested that MySQL will sort correctly even if the position is different in length. 

As the emitted timestamp will always be changing between runs and the composite cursor value will be identical for all records in the same sync, getting into a bad state is very unlikely. 

![image](https://github.com/airbytehq/airbyte/assets/24193788/7137acea-774f-41f5-9d6e-af6b832c228b)
![image](https://github.com/airbytehq/airbyte/assets/24193788/4630d381-d0b3-4ed4-b417-513e680e6de8)

Reply:
.... if we don't expect this case to be relevant, then can we remove the log pos entirely?

but if we do want to keep the log pos... what happens if you use these two cursor values?
```
2023-01-01T00:00:00Z_mysql-bin.1234_123
2023-01-01T00:00:00Z_mysql-bin.1234_4
```
Is there a maximum length for the log pos number? e.g. if it's an int32, can we just `leftpad(pos, '0', 10)`, since int32_max_value = 2.3 billion?

Reply:
We can not remove the log position. You could have 2 events (INSERT, UPDATE) to the same primary key in the same log file. And Airbyte might see both these two records in the same sync giving them the same emittedAt value. Thus the cursor for these two records would be exact same. That's why we need the position. 

But I am not fully convinced about the scenario where the length of the cursor doesnt match. Will all the destinations behave the same if the cursor is of different length? How would Snowflake and Bigquery behave? Will the sort order be correct?

Reply:
The more I think of this, the more this gets dependent on the destination's sort order. And I dont want to be in a situation where two destinations behave differently. 
What if we dont use the file and log position and just use a monotonically increasing value like a number. We take the `emittedAt` value, convert it to epoch seconds or milliseconds and for each record that we see, we keep adding 1 to the original value and use that value as a cursor. Since debezium guarantees that the order of records that it returns to be correct, this will always increase in the right way. Also for records read during snapshot we  should make sure that the cursor value for all records stay the same. The addition only happens for records read from binlogs.

Reply:
What do others think of this idea?^

Reply:
This would also save us in a scenario where the customer changes the binlog file name format between syncs and would be more durable behaviour.

Reply:
Implemented!